common/authorization/interceptor.go

```diff
@@ -45,42 +50,82 @@ func (a *interceptor) Interceptor(
 	handler grpc.UnaryHandler,
 ) (interface{}, error) {
 
-	var namespace string
-	requestWithNamespace, ok := req.(requestWithNamespace)
-	if ok {
-		namespace = requestWithNamespace.GetNamespace()
-	}
+	var claims *Claims
 
-	apiName := info.FullMethod
-	index := strings.LastIndex(apiName, "/")
-	if index > -1 {
-		apiName = apiName[index+1:]
+	if a.claimMapper != nil && a.authorizer != nil {
+		var tlsSubject *pkix.Name
+		var authHeaders []string
+
+		if md, ok := metadata.FromIncomingContext(ctx); ok {
+			authHeaders = md["authorization"]
+		}
+		if p, ok := peer.FromContext(ctx); ok {
+			if tlsAuth, ok := p.AuthInfo.(credentials.TLSInfo); ok {
+				if len(tlsAuth.State.VerifiedChains) > 0 && len(tlsAuth.State.VerifiedChains[0]) > 0 {
+					tlsSubject = &tlsAuth.State.VerifiedChains[0][0].Subject
+				}
+			}
+		}
+		// Add auth into to ctx only if there's some auth info
+		if tlsSubject != nil || len(authHeaders) > 0 {
+			var authHeader string
+			if len(authHeaders) > 0 {
+				authHeader = authHeaders[0]
+			}
+			authInfo := AuthInfo{authHeader, tlsSubject}
+			mappedClaims, err := a.claimMapper.GetClaims(&authInfo)
+			if err != nil {
+				return nil, err
+			}
+			claims = mappedClaims
+			ctx = context.WithValue(ctx, "auth-mappedClaims", mappedClaims)
+		}
 	}
 
-	scope := a.getMetricsScope(metrics.NumAuthorizationScopes, namespace)
-	sw := scope.StartTimer(metrics.ServiceAuthorizationLatency)
-	defer sw.Stop()
+	if a.authorizer != nil {
```

Comment:
Is it a valid scenario to have a non-nil authorize, but a nil claims mapper? Should we just fail on startup if there aren't any?

Reply:
I think this is a corner case but valid scenario. Authorizer can make allow/deny decisions based on some other pieces of data, for example, context. So I think we should allow it.